src/main/java/frc/robot/scoreStates/AmpState.java

```diff
@@ -8,49 +8,88 @@
 import edu.wpi.first.wpilibj2.command.Command;
 import edu.wpi.first.wpilibj2.command.Commands;
 import frc.robot.Constants;
+import frc.robot.subsystems.elevator.Elevator;
+import frc.robot.subsystems.elevator.ElevatorConstants;
+import frc.robot.subsystems.gripper.Gripper;
+import frc.robot.subsystems.gripper.GripperConstants;
 import frc.robot.swerve.SwerveDrive;
 import java.util.Set;
 import lib.Utils;
 
 public class AmpState implements ScoreState {
+    private Elevator elevator;
+    private Gripper gripper;
+    private boolean isAmpingForward = true;
+    private boolean hasTimeToTurnGripper;
+    private Translation2d ampPose;
+    private double distance;
+    private double robotRotation;
+    private Rotation2d ampRotation;
+
+    public AmpState() {
+        elevator = Elevator.getInstance();
+        gripper = Gripper.getInstance();
+    }
 
     @Override
-    public Command driveToClosestOptimalPoint() {
-        return Commands.defer(
+    public Command initializeCommand() {
+        return Commands.runOnce(
```

Comment:
This needs to stay a `DeferredCommand`. Otherwise, everything here will be evaluated once. I would go an extra step and change the return type of this function to `DeferredCommand`.

Reply:
The initialize command only changes the members of the class. This happens in a runnable that the command runs. It doesn't happen during compilation.

Reply:
One of the things done in this command is checking our alliance's color. This data might not be available when this command is created (usually the roboRIO boots faster than the radio, so a connection to the FMS is not yet available). [This relevant Chief Delphi thread](https://www.chiefdelphi.com/t/getalliance-always-returning-red/425782?u=dan) was quite interesting.

Reply:
This data is updated every time the function is called:
```
default boolean isRed() {
        var alliance = DriverStation.getAlliance();
        return alliance.isPresent() && alliance.get() == DriverStation.Alliance.Red;
    }
```
Notice that every call we do DriverStation.getAlliance();

Reply:
But because this check happens in a `Runnable` it is already deferred.

Reply:
It depends on how you'll use these functions in the code. Unless the will be created at runtime, the problem I raised will happen.

Reply:
Yes, that makes sense. Great explanation, thanks! Only thing is that I guess you mean `CommandScheduler` instead of `CommandBase`.
So i see only two required changes:
1. Use `runOnce()` instead of `run()`.
2. Line 68 needs to be moved inside the runnable.

Reply:
It is runOnce

Reply:
Didn't notice I was looking at the shoot state code. Resolving this and opening there.